File: mobile/android/app/src/main/kotlin/ai/pazl/alma/ui/screens/SystemsScreen.kt
package ai.pazl.alma.ui.screens

import ai.pazl.alma.R
import ai.pazl.alma.core.AppContainer
import ai.pazl.alma.core.ScreenState
import ai.pazl.alma.core.SessionHolder
import ai.pazl.alma.data.AlmaClient
import ai.pazl.alma.data.AlmaSystem
import ai.pazl.alma.data.ApiResult
import ai.pazl.alma.data.dto.CalcRequest
import ai.pazl.alma.data.dto.HubEntryDto
import ai.pazl.alma.ui.components.QuietButton
import ai.pazl.alma.ui.components.StateHost
import ai.pazl.alma.ui.sky.NightSky
import ai.pazl.alma.ui.sky.SkyConfig
import ai.pazl.alma.ui.theme.AlmaPalette
import ai.pazl.alma.ui.theme.AlmaTheme
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlinx.serialization.json.JsonObject

/**
 * The eight, grouped by the question they answer rather than by the name of the
 * tradition.
 *
 * A newcomer does not know what a solar return is, but does know "what happens
 * this year". Every group is listed, because which of them has anything in it is
 * the *hub's* decision: give a person a birth time and "this year" stops being
 * empty, and a ready system with no heading to sit under would simply vanish.
 *
 * Two requests. The hub is the screen. The natal chart is only the portrait
 * strip at the top, and it is skipped entirely until there is birth data —
 * firing it regardless meant one guaranteed-failing request on every visit by
 * somebody who has not entered a date yet.
 */
@Composable
fun SystemsScreen(
    container: AppContainer,
    onOpenSystem: (String) -> Unit,
    onAddPerson: () -> Unit,
    onAddBirthData: () -> Unit,
) {
    val vm: SystemsViewModel = viewModel { SystemsViewModel(container.client, container.session) }
    val state by vm.state.collectAsStateWithLifecycle()

    NightSky(config = SkyConfig(seed = 2, motes = 2)) {
        StateHost(state, onRetry = vm::reload) { systems ->
            SystemsBody(
                systems,
                onOpenSystem = onOpenSystem,
                onAddPerson = onAddPerson,
                onAddBirthData = onAddBirthData,
            )
        }
    }
}

/* ── what the screen holds ─────────────────────────────────────────────── */

@Immutable
data class SystemsView(
    val hasBirthData: Boolean,
    /** The hub's own list, in the hub's own order, joined to nothing. */
    val entries: List<HubEntryDto>,
    /**
     * The natal preview, for the three pills at the top.
     *
     * There is deliberately no fallback behind it. A plausible-looking Sun sign
     * that belongs to nobody would be a lie about the one number most people
     * came to see, so a failed chart shows no pills at all.
     */
    val chart: JsonObject?,
)

class SystemsViewModel(
    private val client: AlmaClient,
    private val session: SessionHolder,
) : ViewModel() {

    private val _state = MutableStateFlow<ScreenState<SystemsView>>(ScreenState.Loading)
    val state: StateFlow<ScreenState<SystemsView>> = _state.asStateFlow()

    private var running: Job? = null

    init {
        reload()
    }

    fun reload() {
        running?.cancel()
        running = viewModelScope.launch {
            _state.value = ScreenState.Loading
            val account = session.state.first { it.ready }

            when (val hub = client.hub()) {
                is ApiResult.Err -> _state.value = ScreenState.Failed(hub.failure)
                is ApiResult.Ok -> {
                    // The hub is the authority on whether there is birth data;
                    // the session's own answer is the same fact one request
                    // earlier, and the two only disagree if a second device
                    // changed something mid-session.
                    val chart = if (hub.data.hasBirthData) {
                        (client.system(AlmaSystem.NATAL, CalcRequest(locale = account.locale))
                            as? ApiResult.Ok)?.data?.data
                    } else {
                        null
                    }
                    _state.value = ScreenState.Loaded(
                        SystemsView(
                            hasBirthData = hub.data.hasBirthData,
                            entries = hub.data.systems,
                            chart = chart,
                        )
                    )
                }
            }
        }
    }
}

/* ── the screen ────────────────────────────────────────────────────────── */

/**
 * "calculated" means unlocked and "open" means computable but unpaid — and the
 * numbers are free either way, so both belong under a question heading. Only the
 * three statuses that name a missing detail fall through to "not yet".
 */
private fun HubEntryDto.isReady(): Boolean = status == "calculated" || status == "open"

@Composable
private fun SystemsBody(
    systems: SystemsView,
    onOpenSystem: (String) -> Unit,
    onAddPerson: () -> Unit,
    onAddBirthData: () -> Unit,
) {
    CabinetPage {
        val ready = systems.entries.count { it.isReady() }
        // "5/8" rather than "5 of 8": the total is however many systems the hub
        // lists, and the only translatable part is the word beside it.
        val tally = if (systems.entries.isEmpty()) {
            null
        } else {
            "$ready/${systems.entries.size} ${stringResource(R.string.cabinet_calculated)}"
        }

        ScreenTitle(stringResource(R.string.nav_systems))
        if (tally != null) {
            Text(
                text = tally,
                style = AlmaTheme.type.meta,
                color = AlmaPalette.Gold,
                modifier = Modifier.padding(top = 4.dp),
            )
        }

        if (!systems.hasBirthData) {
            // Nothing here can be computed without a birth date. The graph gives
            // this screen no route to the form — `TodayScreen` is the one that
            // carries `onAddBirthData` — so it says the sentence rather than
            // drawing a button that cannot lead anywhere.
            // A button, not just a sentence. This said "Add your birth date and
            // I can read you." with nothing to tap, because the graph handed
            // this screen no route to the form — `TodayScreen` was the only one
            // given `onAddBirthData`. The state is nearly unreachable (the start
            // destination sends anybody without a profile into the journey) and
            // it was a dead end whenever it did happen.
            NeedBirthData {
                QuietButton(
                    text = stringResource(R.string.state_add_birth_data),
                    onClick = onAddBirthData,
                )
            }
            return@CabinetPage
        }

        Spacer(Modifier.height(22.dp))
        Portrait(systems.chart)
        Spacer(Modifier.height(10.dp))

        // Synthesis has its own block at the bottom and never doubles up here.
        val listed = systems.entries.filter { it.slug != AlmaSystem.SYNTHESIS }

        GroupOrder.forEach { group ->
            val rows = listed.filter { SystemGroups[it.slug] == group && it.isReady() }
            if (rows.isEmpty()) return@forEach
            Spacer(Modifier.height(22.dp))
            RuledLabel(stringResource(group))
            rows.forEachIndexed { index, entry ->
                SystemRow(entry, last = index == rows.lastIndex, onOpenSystem, onAddPerson)
            }
        }

        val pending = listed.filterNot { it.isReady() }
        if (pending.isNotEmpty()) {
            Spacer(Modifier.height(22.dp))
            // One heading covers a missing birth time, a missing person and a
            // system not reached yet alike — "not yet" is the one label true of
            // all three.
            RuledLabel(stringResource(R.string.cabinet_not_yet))
            pending.forEachIndexed { index, entry ->
                SystemRow(entry, last = index == pending.lastIndex, onOpenSystem, onAddPerson)
            }
        }

        systems.entries.firstOrNull { it.slug == AlmaSystem.SYNTHESIS }?.let { synthesis ->
            Spacer(Modifier.height(22.dp))
            RuledLabel(stringResource(R.string.group_all_of_it))
            CabinetRow(onClick = { onOpenSystem(AlmaSystem.SYNTHESIS) }, rule = false) {
                Column(Modifier.weight(1f)) {
                    Text(text = systemName(synthesis.slug), style = AlmaTheme.type.headingM)
                }
                Text(text = "→", style = AlmaTheme.type.positions)
            }
        }

        Spacer(Modifier.height(26.dp))
        Text(text = stringResource(R.string.cabinet_eight_tail), style = AlmaTheme.type.meta)
        Spacer(Modifier.height(24.dp))
    }
}

/**
 * Where a row goes.
 *
 * Almost always to its own system screen. The exception is the status that names
 * a missing *person* rather than a missing detail of the chart: compatibility
 * with nobody to compare against would open a screen whose only content is "add
 * somebody", so the row goes straight to the place that adds one.
 */
@Composable
private fun SystemRow(
    entry: HubEntryDto,
    last: Boolean,
    onOpenSystem: (String) -> Unit,
    onAddPerson: () -> Unit,
) {
    val ready = entry.isReady()
    CabinetRow(
        rule = !last,
        onClick = {
            if (entry.status == "add-person") onAddPerson() else onOpenSystem(entry.slug)
        },
    ) {
        Text(
            text = systemName(entry.slug),
            style = AlmaTheme.type.headingM,
            color = if (ready) AlmaPalette.InkLight else AlmaPalette.Muted,
            modifier = Modifier.weight(1f),
        )
        // The hub carries a slug and a status and nothing else. A second line
        // here — "life path 7", "3 active today" — would be a fact about the
        // person that no request on this screen makes.
        StatusTag(statusLabel(entry.status), ready)
    }
}

@Composable
private fun statusLabel(status: String): String = when (status) {
    "calculated" -> stringResource(R.string.cabinet_calculated)
    "open" -> stringResource(R.string.cabinet_open)
    "needs-time" -> stringResource(R.string.cabinet_needs_time)
    "add-person" -> stringResource(R.string.cabinet_add_person)
    "not-yet" -> stringResource(R.string.cabinet_not_yet)
    // A status this build has not heard of. The hub is the backend's list and
    // it can grow; printing it is more use than printing nothing.
    else -> status
}

/**
 * The portrait: the ring, the sun's glyph inside it, and the three placements a
 * locked chart is still allowed to name.
 *
 * "☉ 23° ♓︎" when the chart is open, "☉ ♓︎" when only the sign names came back.
 * No sign, no pill — the ring stays bare rather than showing a placement nobody
 * calculated.
 */
@Composable
private fun Portrait(chart: JsonObject?) {
    val placements = chart?.obj("placements")
    val sunGlyph = placements?.obj("sun")?.text("sign_glyph")
        ?: chart?.text("sun_sign")?.let { SignGlyphs[it] }

    // Rows, not capsules — the owner's verdict on the stacked ovals was
    // aesthetic and final. The same three facts, printed the way the rest of
    // the product prints a citation: name quiet on the left, position in the
    // serif on the right, a hairline between. Mirrors the iOS front page.
    val rows = listOfNotNull(
        portraitRow(stringResource(R.string.cabinet_sun), chart?.text("sun_sign"), placements?.obj("sun")),
        portraitRow(stringResource(R.string.cabinet_moon), chart?.text("moon_sign"), placements?.obj("moon")),
        chart?.obj("angles")?.obj("formatted")?.text("ascendant")
            ?.let { stringResource(R.string.cabinet_ascendant) to spellSigns(it) }
            ?: chart?.text("rising_sign")?.let {
                stringResource(R.string.cabinet_ascendant) to signWord(it)
            },
    )

    Column(Modifier.fillMaxWidth()) {
        Box(Modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
            ZodiacRing(diameter = 150.dp) {
                if (sunGlyph != null) {
                    Text(
                        text = sunGlyph,
                        style = AlmaTheme.type.displayXl,
                        color = AlmaPalette.StarFill,
                    )
                }
            }
        }
        rows.forEachIndexed { index, (label, value) ->
            CabinetRow(rule = index < rows.lastIndex) {
                Text(text = label, style = AlmaTheme.type.meta, modifier = Modifier.weight(1f))
                Text(text = value, style = AlmaTheme.type.positions)
            }
        }
    }
}

@Composable
private fun portraitRow(
    label: String,
    sign: String?,
    placement: JsonObject?,
): Pair<String, String>? {
    val formatted = placement?.text("formatted")
    if (formatted != null) {
        val house = placement.int("house")?.let { " · " + houseWord(it) }.orEmpty()
        return label to spellSigns(formatted) + house
    }
    return sign?.let { label to signWord(it) }
}

File: mobile/ios/Alma/Screens/Systems/SystemsScreen.swift
import SwiftUI

/// Tab 2 — the hub: all eight systems and the state each is in.
///
/// Two decisions carried over from `src/app/(cabinet)/systems/page.tsx`, both
/// worth keeping.
///
/// **Grouped by the person's question, not by the name of the tradition.**
/// Nobody arrives knowing what a solar return is; everybody arrives wanting to
/// know about this year. The grouping is design metadata (`SystemSlug.group`) —
/// the hub sends a slug and a status and nothing else.
///
/// **Readiness comes before payment.** "calculated" means unlocked and "open"
/// means computable but not paid for, and *both* are ready, because every
/// calculation is free forever and only the writing is sold. Only the three
/// statuses that name something missing — a birth time, a second person, a
/// system not reached yet — fall through to the last group.
struct SystemsScreen: View {

    @Environment(AlmaSessionModel.self) private var session
    @Environment(AppRouter.self) private var router

    @State private var portrait: ScreenState<CalcResult> = .loading

    var body: some View {
        ScreenScaffold(eyebrow: nil, title: nil, seed: 0x5359_5354) {
            header
                .riseIn(0)

            // The order of these three matters. "Has this person given us a
            // birth date" is asked first, off the session rather than off the
            // hub, because a hub that failed to load on a fresh install used to
            // put an offline failure in front of somebody who has nothing to
            // load yet — including a reviewer on a bad connection, on the screen
            // that is supposed to make the case for the product.
            if !session.hasBirthData {
                EmptyArgument { router.openJourney() }
            } else if let hub = session.hub, hub.hasBirthData {
                pills
                    .riseIn(1)
                groups(hub)
                synthesisBlock
                    .riseIn(6)
            } else {
                // There is a birth date and the hub does not know about it: that
                // load failed. Said, with a way to ask again — not an empty list
                // of systems that reads as "you have none".
                AlmaFailure(error: .offline) { await session.refreshHub() }
                    .frame(minHeight: 220)
            }
        }
        // Keyed, for the reason `TodayScreen` gives at length: the tab stacks
        // all stay alive, so a bare `.task` runs once at launch — before the
        // journey — and never again.
        .task(id: session.profile?.id) {
            // Skipped entirely until there is birth data: firing it regardless
            // is one guaranteed-failing request on every visit by somebody who
            // has not entered a date yet.
            guard session.hasBirthData, portrait.value == nil else { return }
            let client = session.client
            let locale = session.locale
            portrait = await almaLoad { try await client.compute(.natal, locale: locale) }.value
        }
    }

    // MARK: — the top of the page

    private var header: some View {
        HStack(alignment: .firstTextBaseline) {
            Text(L10n.tabSystems).almaDisplayL()
            Spacer(minLength: 12)
            if let tally {
                Text(verbatim: tally)
                    .font(.almaNumeral)
                    .foregroundStyle(Color.almaGold)
            }
        }
    }

    /// "5/8", plus the word for it. Nothing at all until the hub answers — a
    /// dangling "calculated" with no number beside it reads as a rendering
    /// fault.
    private var tally: String? {
        guard let hub = session.hub else { return nil }
        let ready = hub.systems.filter { HubStatus.isReady($0.status) }.count
        return "\(ready)/\(hub.systems.count) \(String(localized: L10nCabinet.calculatedWord))"
    }

    /// The portrait: the three placements everybody came to see.
    ///
    /// There is deliberately no fallback behind it. A locked natal answer still
    /// names the three signs, which is what these pills show; a wrong sun sign
    /// on the front page would be a lie about the one number the person came
    /// for, so a failed chart draws nothing rather than something plausible.
    @ViewBuilder
    private var pills: some View {
        // Rows, not capsules. The owner's verdict on the stacked ovals was
        // aesthetic and final; the same three facts read better the way the
        // rest of the product prints a citation — the name quiet on the left,
        // the position in the serif on the right, a hairline between.
        let rows = portraitRows
        if !rows.isEmpty {
            VStack(spacing: 0) {
                ForEach(Array(rows.enumerated()), id: \.offset) { index, row in
                    FactRow(label: row.0, value: row.1, isPosition: true)
                }
            }
            .padding(.top, 2)
        }
    }

    private var portraitRows: [(String, String)] {
        guard let chart = portrait.value?.data else { return [] }
        func placement(_ body: String, fallbackSign: String) -> (String, String)? {
            let label = L10nCabinet.bodyName(body)
            if let raw = chart["placements"]?[body]?["formatted"]?.stringValue {
                let house = chart["placements"]?[body]?["house"]?.intValue
                    .map { " · " + L10nCabinet.houseName($0) } ?? ""
                return (label, ChartFacts.spellSigns(in: raw) + house)
            }
            if let sign = chart[fallbackSign]?.stringValue {
                return (label, L10nCabinet.signName(sign))
            }
            return nil
        }
        var rows: [(String, String)] = []
        if let sun = placement("sun", fallbackSign: "sun_sign") { rows.append(sun) }
        if let moon = placement("moon", fallbackSign: "moon_sign") { rows.append(moon) }
        if let raw = chart["angles"]?["formatted"]?["ascendant"]?.stringValue {
            rows.append((String(localized: L10nCabinet.ascendant), ChartFacts.spellSigns(in: raw)))
        } else if let sign = chart["rising_sign"]?.stringValue {
            rows.append((String(localized: L10nCabinet.ascendant), L10nCabinet.signName(sign)))
        }
        return rows
    }

    // MARK: — the eight

    @ViewBuilder
    private func groups(_ hub: Hub) -> some View {
        let entries = hub.systems.filter { $0.slug != .synthesis }
        let ready = entries.filter { HubStatus.isReady($0.status) }
        let pending = entries.filter { !HubStatus.isReady($0.status) }

        ForEach(Array(SystemGroup.allCases.enumerated()), id: \.element) { position, group in
            let rows = ready.filter { $0.slug.group == group }
            if !rows.isEmpty {
                CabinetSection(label: group.title) {
                    ForEach(rows) { entry in row(entry) }
                }
                .riseIn(2 + position)
            }
        }

        if !pending.isEmpty {
            // One heading covers a missing birth time, a missing person and a
            // system not reached yet alike: "not yet" is the one label true of
            // all three.
            CabinetSection(label: L10nCabinet.statusNotYet) {
                ForEach(pending) { entry in row(entry) }
            }
            .riseIn(5)
        }
    }

    private func row(_ entry: HubEntry) -> some View {
        SystemRow(system: entry.slug, status: entry.status) {
            // The one status that names a missing *person* rather than a
            // missing detail of the chart goes straight to the place that adds
            // one: a compatibility screen with nobody to compare against would
            // have "add somebody" as its only content.
            if entry.status == "add-person" {
                router.push(.people)
            } else {
                router.push(.system(entry.slug))
            }
        }
    }

    private var synthesisBlock: some View {
        CabinetSection(label: L10nCabinet.groupAllOfIt) {
            SystemRow(system: .synthesis, status: statusOfSynthesis) {
                router.push(.system(.synthesis))
            }
            // «Rebuilds itself when a system is added» stood here and said
            // nothing a person shopping for a reading needed — the owner cut it.
        }
    }

    private var statusOfSynthesis: String {
        session.status(of: .synthesis)?.status ?? "not-yet"
    }
}

#Preview {
    SystemsScreen()
        .environment(AlmaSessionModel.preview())
        .environment(AppRouter())
}
